### PR TITLE
KAFKA-14639: A single partition may be revoked and assign during a single round of rebalance

### PR DESCRIPTION
**Pull request by philipnee**
This is a really long story, but the incident started in KAFKA-13419 when we observed a member sending out a topic partition owned from the previous generation when a member missed a rebalance cycle due to REBALANCE_IN_PROGRESS.

After a long debate, we've decided the best thing to do at the moment is to continue to honor the partition owned by members with the highest generation.

In this PR, I changed the AbstractStickyAssignor.AllSubscriptionsEqual method.  In short, it should no long check and validate only the highest generation.  Instead, we consider 3 cases:
1. Member will continue to hold on to its partition if there are no other owners
2. If there are 1+ owners to the same partition. One with the highest generation will win.
3. If two members of the same generation hold on to the same partition.  We will log an error but remove both from the assignment. (Same with the current logic)

Here are some important notes that lead to the PR:
- If a member is kicked out of the group, and `UNKNOWN_MEMBER_ID` will be thrown.
- It seems to be a common situation that members are late to joinGroup and therefore get `REBALANCE_IN_PROGRESS` error.  This is why we don't want to reset generation because it might cause lots of revocations and can be disruptive

To summarize the current behavior of different errors:
`REBALANCE_IN_PROGRESS`
- heartbeat: requestRejoin if member state is stable
- joinGroup: rejoin immediately
- syncGroup: rejoin immediately
- commit: requestRejoin and fail the commit. Raise this exception if the generation is staled, i.e. another rebalance is already in progress.

`UNKNOWN_MEMBER_ID`
- heartbeat: resetStateAndRejoinif generation hasn't changed. otherwise, ignore
- joinGroup: resetStateAndRejoin if generation unchanged, otherwise rejoin immediately
- syncGroup:  resetStateAndRejoin if generation unchanged, otherwise rejoin immediately

`ILLEGAL_GENERATION`
- heartbeat: resetStateAndRejoinif generation hasn't changed. otherwise, ignore
- syncGroup: raised the exception if generation has been resetted or the member hasn't completed rebalancing.  then resetStateAndRejoin if generation unchanged, otherwise rejoin immediately

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 1]
```diff
@@ -16,6 +16,13 @@
  */
 package org.apache.kafka.clients.consumer.internals;
 
+import org.apache.kafka.clients.consumer.internals.Utils.PartitionComparator;
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: this is "optimized" by the IDE

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 2]
```diff
@@ -153,43 +158,46 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
...
             consumerToOwnedPartitions.put(consumer, ownedPartitions);
 
-            // Only consider this consumer's owned partitions as valid if it is a member of the current highest
-            // generation, or it's generation is not present but we have not seen any known generation so far
-            if (memberData.generation.isPresent() && memberData.generation.get() >= maxGeneration
+            if (memberData.generation.isPresent() && memberData.generation.get() >= maxGeneration - 1
                 || !memberData.generation.isPresent() && maxGeneration == DEFAULT_GENERATION) {
 
-                // If the current member's generation is higher, all the previously owned partitions are invalid
-                if (memberData.generation.isPresent() && memberData.generation.get() > maxGeneration) {
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java:162]: we don't need this because the maxGeneration has already been computed previously, so we know the set of valid members.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 3]
```diff
@@ -153,43 +158,45 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
...
+                                        consumer, otherConsumer, tp, maxGeneration);
+                                    partitionsWithMultiplePreviousOwners.add(tp);
+                                }
+                                consumerToOwnedPartitions.get(otherConsumer).remove(tp);
+                                allPreviousPartitionsToOwner.put(tp, consumer);
+                                continue;
+                            }
+
+                            if (memberGeneration > otherMemberGeneration) {
+                                log.warn("Found multiple consumers {} and {} claiming the same TopicPartition {} in " +
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: I feel... duplicated ownership should be a warning.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 4]
```diff
@@ -835,6 +835,7 @@ public void handle(SyncGroupResponse syncResponse,
                 } else if (error == Errors.REBALANCE_IN_PROGRESS) {
                     log.info("SyncGroup failed: The group began another rebalance. Need to re-join the group. " +
                                  "Sent generation was {}", sentGeneration);
+                    resetStateAndGeneration("member missed the rebalance", true);
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: @philipnee I am trying to convince myself about this change. This basically means that whenever a member is late for the sync-group phase, it will abandon all its partitions. Here late means that the member sends the sync-group request after the next rebalance has started. I wonder how common this is, especially in large groups.

My understanding is that all pending sync-group requests are completed when the leader sends the assignment. When they are completed, the members with partitions to be revoked calls revoke them and re-join more-or-less immediately (because we don't commit offsets in the cooperative mode, I think). This makes me think that this will happen regularly in medium to large groups.

Could you elaborate a bit more on the reasoning behind this conservative change?

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Hey @dajac  thanks for the review. I think in case of the member misses a generation, we want to make sure the owned partitions are revoked (due to generation reset).  Regardlessly, it should still rejoin with its current partitions and should continue to hold on to its partition if it is only 1 generation behind. If it is 1+ generations behind, circle back to the beginning of my response, we want to make sure they are revoked because the partition might have already been reassigned.

This makes me think that this will happen regularly in medium to large groups. -> I think this might not be as uncommon as what we think, especially with a large consumer group deployed to multiple pods, considering the pods can be staled before sending out syncGroup, while another consumer in a different pod tries to join the group.

I hope i'm answering your questions there, I apologize if I misunderstood anything.

Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: > Regardlessly, it should still rejoin with its current partitions and should continue to hold on to its partition if it is only 1 generation behind.

Are you sure about this? It seems to me that whenever the generation is reset, all the partitions are lost and the member rejoins with no owned partitions. Am I wrong?

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Hmm, I think the partitions are only lost during the onJoinPrepare, what I'm thinking is this: https://github.com/apache/kafka/blob/61530d68ce83467de6190a52da37b3c0af84f0ef/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java#L507-L511

The gist is: on one of the 4 exception thrown in join/sync group, it should immediately re-send the join request. Are you thinking about the how client handles the illegal generation error? I think it is only being thrown during sync group and heartbeat. So I think by just resetting the generation shouldn't immediately causes revocation.

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Added to my previous comment, I think you will need to set the `needsOnJoinPrepare` to true to go through the revocation as pointed out here: https://github.com/apache/kafka/blob/61530d68ce83467de6190a52da37b3c0af84f0ef/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L821

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Offline discussion with @hachikuji  - it seems like what we want is to revoke the old partitions but resend these partitions on the subsequent join. RN, join only sends out the assigned partitions.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 5]
```diff
@@ -835,6 +839,8 @@ public void handle(SyncGroupResponse syncResponse,
                 } else if (error == Errors.REBALANCE_IN_PROGRESS) {
                     log.info("SyncGroup failed: The group began another rebalance. Need to re-join the group. " +
                                  "Sent generation was {}", sentGeneration);
+                    savePartitionAndGenerationState();
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: saving the generation and partition state prior to resetting it.

Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: couldn’t we just rejoin without reseting the generation? in other words, what is the advantage of resetting the generation here if we still want to use it to re-join afterwards?

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: i think the ration for resetting the generation is to ensure member's partitions get revoke during the onJoinPrepare, i'm trying to think is there's a better way to do this...

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: I'm not sure what's the better way to invoke `onPartitionsLost` w/o resetting the generation. The alternative is to use a flag, but that's rather unclean.

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: This is discarded in the latest commit

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 4 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: this is discarded in the latest commit

**Comment by philipnee** [comment]
Hey @mimaison - Sorry for the late notice. We've been trying to fix this issue for the coming release 3.5.  Would it be possible to include this for the 3.5 release? We should be able to finish reviewing the changes this week. cc @dajac

**Comment by mimaison** [comment]
@philipnee Yes, we can backport this to 3.5

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 6]
```diff
@@ -137,7 +136,7 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
         Map<TopicPartition, String> allPreviousPartitionsToOwner = new HashMap<>();
 
         for (Map.Entry<String, Subscription> subscriptionEntry : subscriptions.entrySet()) {
-            String consumer = subscriptionEntry.getKey();
+            final String consumer = subscriptionEntry.getKey();
             Subscription subscription = subscriptionEntry.getValue();
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: nit: While here, should this one be final as well?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 7]
```diff
@@ -149,47 +148,51 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
             }
 
             MemberData memberData = memberData(subscription);
+            maxGeneration = Math.max(maxGeneration, memberData.generation.orElse(DEFAULT_GENERATION));
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: nit: Should we define a variable for `memberData.generation.orElse(DEFAULT_GENERATION)`? The same code is reused later.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 8]
```diff
@@ -149,47 +148,51 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
...
+                    String otherConsumer = allPreviousPartitionsToOwner.put(tp, consumer);
+                    if (otherConsumer == null) {
+                        // this partition is not owned by other consumer in the same generation
+                        ownedPartitions.add(tp);
+                    } else {
+                        final int memberGeneration = memberData.generation.orElse(DEFAULT_GENERATION);
+                        final int otherMemberGeneration = subscriptions.get(otherConsumer).generationId().orElse(DEFAULT_GENERATION);
+
+                        if (memberGeneration == otherMemberGeneration) {
+                            if (subscriptions.get(otherConsumer).generationId().orElse(DEFAULT_GENERATION) == memberData.generation.orElse(DEFAULT_GENERATION)) {
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Aren't those two if statements the same?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 9]
```diff
@@ -149,47 +148,51 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
...
+                                log.error("Found multiple consumers {} and {} claiming the same TopicPartition {} in the "
+                                        + "same generation {}, this will be invalidated and removed from their previous assignment.",
+                                    consumer, otherConsumer, tp, memberGeneration);
+                                partitionsWithMultiplePreviousOwners.add(tp);
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: So my understanding is that partitions put in `partitionsWithMultiplePreviousOwners` will be unassigned from all consumers claiming them. Is my understanding correct?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 10]
```diff
@@ -149,47 +148,51 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
...
+                            }
+                            consumerToOwnedPartitions.get(otherConsumer).remove(tp);
+                            allPreviousPartitionsToOwner.put(tp, consumer);
+                            continue;
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: This does not look nice. Should we remove it and use `else if`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 11]
```diff
@@ -149,47 +148,51 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
...
+                            consumerToOwnedPartitions.get(consumer).add(tp);
                             consumerToOwnedPartitions.get(otherConsumer).remove(tp);
-                            partitionsWithMultiplePreviousOwners.add(tp);
+                            allPreviousPartitionsToOwner.put(tp, consumer);
                         }
+
+                        // if memberGeneration < otherMemberGeneration, the other member continue owns the generation
+                        log.warn("Found multiple members {} and {} claiming the same TopicPartition {} in " +
+                                "different generations. The topic partition wil be assigned to the member with " +
+                                "the higher generation {}.",
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Should we add the generation for each members? This will also tell us which member got it in the end. I would also use `consumers` instead of `members` in order to be consistent with the other logs.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java [hunk 12]
```diff
@@ -1038,6 +1038,96 @@ public void testPartitionsTransferringOwnershipIncludeThePartitionClaimedByMulti
         assertTrue(isFullyBalanced(assignment));
     }
 
+    @ParameterizedTest(name = TEST_NAME_WITH_RACK_CONFIG)
+    @EnumSource(RackConfig.class)
+    public void testOwnedPartitionsAreStableForConsumerWithMultipleGeneration(RackConfig rackConfig) {
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: nit: It seems that we also test the case where a partition is claimed by two consumers with different generation in this test. Should we try to reflect this in its name?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java [hunk 13]
```diff
@@ -1038,6 +1038,96 @@ public void testPartitionsTransferringOwnershipIncludeThePartitionClaimedByMulti
...
+            partitions(tp(topic3, 2), tp(topic2, 2), tp(topic1, 1)), currentGeneration - 2, 2));
+        subscriptions.put(consumer4, buildSubscriptionV2Above(topics(topic, topic2, topic3, topic1),
+            partitions(tp(topic3, 1), tp(topic, 1), tp(topic, 2)), currentGeneration - 3, 3));
+
+        Map<String, List<TopicPartition>> assignment = assignor.assignPartitions(partitionsPerTopic, subscriptions);
+        // ensure assigned partitions don't get reassigned
+        assertTrue(assignment.get(consumer1).containsAll(
+            Arrays.asList(tp(topic2, 1),
+                tp(topic3, 0),
+                tp(topic1, 2))));
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: nit: It may be better to verify the full assignment for completeness.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java (hunk 12 above)
Comment (dajac) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: I actually wonder if this was expected or if it is just a typo in the test case because there is also `testOwnedPartitionsAreInvalidatedForConsumerWithMultipleGeneration`.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java [hunk 14]
```diff
@@ -1038,6 +1038,96 @@ public void testPartitionsTransferringOwnershipIncludeThePartitionClaimedByMulti
...
+        assertEquals(new HashSet<>(partitions(tp(topic, 0), tp(topic2, 1), tp(topic, 1))),
+            new HashSet<>(assignment.get(consumer1)));
+        assertEquals(new HashSet<>(partitions(tp(topic, 2), tp(topic2, 2), tp(topic2, 0))),
+            new HashSet<>(assignment.get(consumer2)));
+        assertTrue(assignor.partitionsTransferringOwnership.isEmpty());
+
+        verifyValidityAndBalance(subscriptions, assignment, partitionsPerTopic);
+        assertTrue(isFullyBalanced(assignment));
+    }
+
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:1129]: Should we also add a test for the case where a partition is claimed by two consumers with the same generation?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java [hunk 15]
```diff
@@ -42,10 +30,22 @@
 import org.junit.jupiter.params.provider.EnumSource;
 import org.junit.jupiter.params.provider.ValueSource;
 
+import java.nio.ByteBuffer;
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: nit: Could we revert this as it is not related to the fix?

**Review by dajac — COMMENTED** [review]
@philipnee Thanks for the update. I made a first pass on it and left a few comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java (hunk 9 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: that seems like the case, reference to the snippet here:
```
for (TopicPartition doublyClaimedPartition : partitionsWithMultiplePreviousOwners) {
                    if (ownedPartitions.contains(doublyClaimedPartition)) {
                        log.error("Found partition {} still claimed as owned by consumer {}, despite being claimed by multiple "
                                        + "consumers already in the same generation. Removing it from the ownedPartitions",
                                doublyClaimedPartition, consumer);
                        ownedPartitions.remove(doublyClaimedPartition);
                    }           
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java (hunk 10 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: It could be.  I got into the habit of returning early, I thought it makes it easier to read.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java (hunk 14 above)
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:1129]: good idea, I didn't check because i thought there could be an existing test case covering it but let me check

Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:1129]: ok `testPartitionsTransferringOwnershipIncludeThePartitionClaimedByMultipleConsumersInSameGeneration` might cover it.  But I'll double check if that's what we want.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java (hunk 15 above)
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: oh yes, will do it.... thanks for my IDE's import optimization.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java (hunk 12 above)
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: Ok I renamed the test to `testEnsurePartitionsAssignedToHighestGeneration` as the goal of this test is to make sure partitions are always assigned to the member with the highest generation.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java (hunk 9 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: From KAFKA-12984: 
```
...the assignor will now explicitly look out for partitions that are being claimed by multiple consumers ... we have to invalidate this partition from the ownedPartitions of both consumers, since we can't tell who, if anyone, has the valid claim to this partition.
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 16]
```diff
@@ -149,47 +148,57 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
...
-                membersOfCurrentHighestGeneration.add(consumer);
-                for (final TopicPartition tp : memberData.partitions) {
-                    // filter out any topics that no longer exist or aren't part of the current subscription
-                    if (allTopics.contains(tp.topic())) {
-                        String otherConsumer = allPreviousPartitionsToOwner.put(tp, consumer);
-                        if (otherConsumer == null) {
-                            // this partition is not owned by other consumer in the same generation
-                            ownedPartitions.add(tp);
-                        } else {
+                        if (memberGeneration == otherMemberGeneration) {
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java:168]: nit: Could we put a comment in this branch like we did for the others?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 17]
```diff
@@ -149,47 +148,57 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
...
-                                     consumer, otherConsumer, tp, maxGeneration);
-                            consumerToOwnedPartitions.get(otherConsumer).remove(tp);
+                                            + "same generation {}, this will be invalidated and removed from their previous assignment.",
+                                    consumer, otherConsumer, tp, memberGeneration);
                             partitionsWithMultiplePreviousOwners.add(tp);
+                            consumerToOwnedPartitions.get(otherConsumer).remove(tp);
+                            allPreviousPartitionsToOwner.put(tp, consumer);
+                        } else if (memberGeneration > otherMemberGeneration) {
+                            // move partition from the member with an older generation to the member with the newer generation
+                            consumerToOwnedPartitions.get(consumer).add(tp);
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: nit: We could use `ownedPartitions.add(tp)`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 18]
```diff
@@ -149,47 +148,57 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
...
+                            consumerToOwnedPartitions.get(otherConsumer).remove(tp);
+                            allPreviousPartitionsToOwner.put(tp, consumer);
+                            // if memberGeneration > otherMemberGeneration, the other member continue owns the generation
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: nit: Should we just remove this one? It seems that the comment earlier explain it all.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 19]
```diff
@@ -149,47 +148,57 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
...
+                            log.warn("{} in generation {} and {} in generation {} claiming the same TopicPartition {} in " +
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: nit: I would say `Consumer {} in generation and consumer {} in ...`. This applies to the other comment as well.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java [hunk 20]
```diff
@@ -149,47 +148,57 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
...
+                                            "different generations. The topic partition wil be assigned to the member with " +
+                                            "the higher generation {}.",
+                                    consumer, memberGeneration,
+                                    otherConsumer, otherMemberGeneration,
+                                    tp,
+                                    memberGeneration);
+                        } else {
+                            // if memberGeneration < otherMemberGeneration, the other member continue owns the generation
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: nit: Should we remove `if memberGeneration < otherMemberGeneration` as it is implicit?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java [hunk 21]
```diff
@@ -1038,6 +1038,99 @@ public void testPartitionsTransferringOwnershipIncludeThePartitionClaimedByMulti
...
+        partitionsPerTopic.put(topic2, partitionInfos(topic2, 3));
+        partitionsPerTopic.put(topic3, partitionInfos(topic3, 3));
+
+        int currentGeneration = 10;
+
+        // ensure partitions are always assigned to the member with the highest generation
+        // topic, 1 -> [consumer2], consumer3
+        // topic2, 1 -> [consumer2], consumer3
+        // topic, 0 -> [consumer1], consumer3
+        // topic3, 2 -> consumer3
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: I don't understand what you are trying to express here. It also seems to me that the mapping is incorrect. Should we just remove this?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java [hunk 22]
```diff
@@ -1038,6 +1038,99 @@ public void testPartitionsTransferringOwnershipIncludeThePartitionClaimedByMulti
...
+
+    @ParameterizedTest(name = TEST_NAME_WITH_RACK_CONFIG)
+    @EnumSource(RackConfig.class)
+    public void testNoReassignmentOnCurrentMembers(RackConfig rackConfig) {
+        initializeRacks(rackConfig);
+        Map<String, List<PartitionInfo>> partitionsPerTopic = new HashMap<>();
+        partitionsPerTopic.put(topic, partitionInfos(topic, 3));
+        partitionsPerTopic.put(topic2, partitionInfos(topic2, 3));
+        partitionsPerTopic.put(topic3, partitionInfos(topic3, 3));
+        partitionsPerTopic.put(topic1, partitionInfos(topic1, 3));
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: nit: Could we put `topic1` after `topic`?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java [hunk 23]
```diff
@@ -1038,6 +1038,99 @@ public void testPartitionsTransferringOwnershipIncludeThePartitionClaimedByMulti
...
+        // topic, 1 -> [consumer2], consumer3
+        // topic2, 1 -> [consumer2], consumer3
+        // topic, 0 -> [consumer1], consumer3
+        // topic3, 2 -> consumer3
+        subscriptions.put(consumer1, buildSubscriptionV2Above(topics(topic, topic2, topic3),
+            partitions(tp(topic, 0), tp(topic2, 0), tp(topic3, 0)), currentGeneration, 0));
+        subscriptions.put(consumer2, buildSubscriptionV2Above(topics(topic, topic2, topic3),
+            partitions(tp(topic, 1), tp(topic2, 1), tp(topic3, 1)), currentGeneration - 1, 1));
+        subscriptions.put(consumer3, buildSubscriptionV2Above(topics(topic, topic2, topic3),
+            partitions(tp(topic3, 0), tp(topic3, 2), tp(topic2, 1)), currentGeneration - 2, 1));
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: nit: Could we put `tp(topic2, 1)` before `topic3`?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java [hunk 24]
```diff
@@ -1038,6 +1038,99 @@ public void testPartitionsTransferringOwnershipIncludeThePartitionClaimedByMulti
...
+        initializeRacks(rackConfig);
+        Map<String, List<PartitionInfo>> partitionsPerTopic = new HashMap<>();
+        partitionsPerTopic.put(topic, partitionInfos(topic, 3));
+        partitionsPerTopic.put(topic2, partitionInfos(topic2, 3));
+        partitionsPerTopic.put(topic3, partitionInfos(topic3, 3));
+        partitionsPerTopic.put(topic1, partitionInfos(topic1, 3));
+
+        int currentGeneration = 10;
+
+        subscriptions.put(consumer1, buildSubscriptionV2Above(topics(topic, topic2, topic3, topic1), partitions(),
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: nit: Could we put `partitions(),` on the next line to follow the pattern used by the others? It is easier to read and compare them.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java [hunk 25]
```diff
@@ -1038,6 +1038,99 @@ public void testPartitionsTransferringOwnershipIncludeThePartitionClaimedByMulti
...
+            partitions(tp(topic, 0), tp(topic2, 0), tp(topic1, 0)), currentGeneration - 1, 1));
+        subscriptions.put(consumer3, buildSubscriptionV2Above(topics(topic, topic2, topic3, topic1),
+            partitions(tp(topic3, 2), tp(topic2, 2), tp(topic1, 1)), currentGeneration - 2, 2));
+        subscriptions.put(consumer4, buildSubscriptionV2Above(topics(topic, topic2, topic3, topic1),
+            partitions(tp(topic3, 1), tp(topic, 1), tp(topic, 2)), currentGeneration - 3, 3));
+
+        Map<String, List<TopicPartition>> assignment = assignor.assignPartitions(partitionsPerTopic, subscriptions);
+        // ensure assigned partitions don't get reassigned
+        assertEquals(new HashSet<>(assignment.get(consumer1)),
+                new HashSet<>(partitions(tp(topic2, 1), tp(topic3, 0), tp(topic1, 2))));
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java]: nit: We should put the expected one first.

**Review by dajac — COMMENTED** [review]
Thanks for the update. I left a few more nits.

From the PR description:
> If two members with the same generation are owning the same partition. We will log an error but remove one from the assignment. (Same with the current logic)

It seems to me that the partition will be removed for all members in this case. Then, depending on the assignor, the partition may be reassigned directly (Sticky) or revoked first (CooperativeSticky).

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java (hunk 20 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: thanks, there's also a typo: continue to own the "topic partition"

**Comment by philipnee** [comment]
@dajac - Thanks again for the review. I've addressed your comments.

**Review by dajac — APPROVED** [review]
LGTM, thanks.

**Comment by dajac** [comment]
Merged to trunk and 3.5. cc @mimaison

**Comment by philipnee** [comment]
Thanks David, FWIW: I think we also want to back port it to 3.4, I'll raise a PR.

**Comment by dajac** [comment]
Merged https://github.com/apache/kafka/pull/13652 to 3.4.
